Block contents icon fixes

jbasiago opened the pull request:
Issues with BlockContentsIcon
	- hard to place icons for blocks that have no value - they don't show up when added - fixed
	- if the block has no value at all, you can't even right click on it to open the editing pane.  You first have to add a value using double click so that you can right click on the icon - fixed
	- sometimes there is no static rendering of rotated icons -- fixed
	- after adding a value to a block, a popup comes up away from the icon, particualarly on non-rotated icons -- fixed
	- when a block value is changed, the icon may be rendered properly (rotated), but then changes to improper rendering -- fixed
	- when background color is added, it's not always rendered properly when icon is rotated -- fixed
	- "hide when empty" does not render correctly initially.  It does hide icons after the value cycles to blank -- fixed
	- after power up, "hide when empty" does not render correctly -- fixed
	- if you go into edit mode on the panel, all of the icons show up, but they don't go away when you go back to non-edit mode -- fixed
	- when a rotated icon with text goes from edited to un-edited, it disappears. -- fixed

danielb987 commented on java/src/jmri/jmrit/display/BlockContentsIcon.java:
```diff
@@ -296,10 +299,6 @@ public void dispose() {
     @Override
     public void doMouseClicked(JmriMouseEvent e) {
         if (e.getClickCount() == 2) { // double click?
-            if (!getEditor().isEditable() && isValueEditDisabled()) {
-                log.debug("Double click block value edit disabled");
-                return;
-            }
```
Why are these lines removed?

jbasiago replied:
That was a mistake.  Those lines should be readded.

Thanks

danielb987 commented:
@dsand47 
Do you have time to look at this PR? I haven't used block icons.

dsand47 commented:
The original changes worked nicely and resolved a number of issues.   However, the latest changes made the empty icon intrusive.

Before:
![Screenshot 2025-04-17 at 6 14 21 PM](https://github.com/user-attachments/assets/d0076f46-d4c5-44af-a71d-0616b69992db)

After:
![Screenshot 2025-04-17 at 6 13 34 PM](https://github.com/user-attachments/assets/9265b09a-c872-4f3a-b731-ec9903218ed3)

jbasiago commented:
I was finding that the empty icon would not receive right clicks to open the editor.  It seemed that you had to have a character in the underlying Label in order to successfully right click, so even when the Block value was empty or null, I inserted a space character into the displayState(key) method (called from displayState() in BlockContentsIcon) to make sure the contents was of type String and to give something to click on.  This caused the rectangle to expand to the height of a character, which seems to be appropriate to me.  I may have misinterpreted this behavior.

dsand47 commented:
> I was finding that the empty icon would not receive right clicks to open the editor. It seemed that you had to have a character in the underlying Label in order to successfully right click, so even when the Block value was empty or null, I inserted a space character into the displayState(key) method (called from displayState() in BlockContentsIcon) to make sure the contents was of type String and to give something to click on. This caused the rectangle to expand to the height of a character, which seems to be appropriate to me. I may have misinterpreted this behavior.

The tricky part of the empty memory and block content icons is having the mouse near the upper left corner.  I don't know what is the actual size of the click detection rectangle.

danielb987 commented:
Would it work to have a small image, about 5 * 5 pixels, with the color of the text background, instead of a space character?

The image could be created on the fly when needed.

dsand47 commented:
> Would it work to have a small image, about 5 * 5 pixels, with the color of the text background, instead of a space character?
> 
> The image could be created on the fly when needed.

The default icon for a null value is 5x5.  The click detection rectangle appears to be about 4x4.

The main problem is that when the block contents label is added to the LE panel, it is invisible.  If you don't move the mouse, you can right click on the icon and set color, borders, height, etc.

jbasiago commented:
So I started with my main fork code again, and added in the changes that I made one at a time to see what was affected.  The first set of changes were to PositionableLabel.  This fixed the emptyHidden behavior when changing from Edit Mode to Display Mode in the layout editor.  The next set of fixes were in the MemoryIcon class to update the display(key) method.  This got the rotation rendering corrected for the most part and still maintained the small 5x5 icon that is desired from above.

At this point, I felt there were two issues that remained.  In edit mode, if you set the BlockContentsIcon to have a fixed width and height plus a non-zero value of rotation, the code will not rotate the rectangle if there is no value in the block.  It seems to have something to do with the icon and no text being non-rotatable.  You get a block that matches the fixed size, but it will not rotate in edit mode.  In non-edit mode, it works fine.  

The second issue that I feel is important is that the small icon (5x5 square) really does not represent the position and rotation of the text when there is no text in the Block.  You have to add text to the block to properly place the icon and see that it is rotated in edit mode.

If you set the empty value of the block to a space, the empty block better represents, in my view, the location, fixed size, and rotation of the label.  Just my opinion.  You guys know this code base and requirements far better than I do.  I'll continue to poke at this to see if I can fix the edit behavior.

Thanks

dsand47 commented:
Other than the remaining change to .idea, the technical changes are good and fix some very old issues.

The remaining issue is that there is a very visible change on the panels.  It is common for people to react negatively to unexpected visual changes.  To address this issue, the release note needs be expanded to describe the visual changes.  

Have the changes been tested using Panel Editor?

danielb987 commented on .idea/ant.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project version="5">
+<project version="4">
```
It would be good to revert this change as well. Thanks!

jbasiago commented:
So I rolled back the remaining ant.xml file.  I also rolled back the changes that inserted a space into empty BlockContentsIcons.  The basic behavior should be very close to the former behavior when editing layout panels except that the rotated blocks render correctly and the emptyHidden flag works.  I have done no testing with the Panel editor.  I didn't realize you guys worked on weekends or I would have gotten these last changes sooner.  What would you like me to do?  It's been great working with you.  Thanks.

jbasiago commented:
I did some basic tests with Panel Editor.  I've never used Panel Editor before.  I set up a panel that mimicked the layout editor panel with the same BlockContentsIcons in similar places.  I configured those icons with rotation, fixed size, emptyHidden, and background color the same as the layout panel test.  Both performed similarly.  The Panel Editor icons never completely hide when empty.  There was a remaining 5x5 block of the background color always remaining.  But the rotations worked when not empty.  The fixed sized icons did not rotate when empty, much like the behavior when in Edit mode in the layout editor.  Any suggestions?

dsand47 commented:
For Panel Editor, unchecking "Panel items popup menus active" switches PE out of edit mode. 

I have not seen the 5x5 icons in PE.  When a block becomes unoccupied, a big red X is displayed.  This is caused because the occupancy sensor going inactive sets the block value to null.  For some reason, LE displays the 5x5 square but PE uses the red X.  If I set the block value to an empty string, PE shows the 5x5 square.

The block value can be null, an empty or non-empty string, or some other object, such as a roster entry or a reporter.  This means the block contents icon has to handle null, an empty string or a non-empty string.  The last one can be text or a "toString" value from an object.

jbasiago commented:
I've looked at this as well.  I'm not sure how to evaluate whether I'm fixing something or breaking it.  For instance, when LE instantiates a BlockContentsIcon, it does not provide a defaultIcon, but when PE does the same, it does provide a defaultIcon.  This is the red cross that appears in a PE panel, but not in an LE panel.  By the way, my experience is that if you save the panel, and restart PanelPro, the red squares go away.  Also, PE panels don't seem to provide the black outline for the 5x5 icon, where as LE panels do.  If you add a background color, you can see the 5x5 icons.  So are you asking me to change whatever behavior is in a PE panel to be like whatever is in an LE panel?  Again, I've never used PE panels.  Sorry for the confusion.  Thanks

dsand47 commented:
> I've looked at this as well. I'm not sure how to evaluate whether I'm fixing something or breaking it. For instance, when LE instantiates a BlockContentsIcon, it does not provide a defaultIcon, but when PE does the same, it does provide a defaultIcon. This is the red cross that appears in a PE panel, but not in an LE panel. By the way, my experience is that if you save the panel, and restart PanelPro, the red squares go away. Also, PE panels don't seem to provide the black outline for the 5x5 icon, where as LE panels do. If you add a background color, you can see the 5x5 icons. So are you asking me to change whatever behavior is in a PE panel to be like whatever is in an LE panel? Again, I've never used PE panels. Sorry for the confusion. Thanks

I don't think anything else needs to be done with PE.  The only difference is the red X which nobody has mentioned.

dsand47 reviewed (approved):
The latest version works nicely.

danielb987 commented:
@jbasiago 
Thanks for your work on this!

danielb987 commented:
The PR is now merged and will be included in the next JMRI test release 5.11.6.
